### PR TITLE
feat(cloud,schemas): add PATCH /tenants/:id API

### PR DESCRIPTION
**Pull request by darcyYe**
## Summary

add PATCH /tenants/:id API

## Testing

UTs and tested locally, can successfully update tenant info.

## Checklist

- [ ] `.changeset`
- [x] unit tests
- [x] integration tests
- [ ] docs
Will add changeset once the whole feature is completed. This modification is transparent to current users.

OR

- [ ] This PR is not applicable for the checklist

### CODE REVIEW

**Review by wangsijie — COMMENTED** [review]
Would you mind updating the PR description? It looks like that the changes are beyond "PATCH /tenants/:id"

**Comment by darcyYe** [comment]
> Would you mind updating the PR description? It looks like that the changes are beyond "PATCH /tenants/:id"

Rebased to hide unrelated changes.

File: packages/cloud/src/routes/tenants.ts [hunk 1]
```diff
@@ -9,6 +9,37 @@ export const tenantsRoutes = (library: TenantsLibrary) =>
...
+      {
+        body: createTenantGuard.pick({ name: true, tag: true }).partial(),
+        response: tenantInfoGuard,
+      },
+      async (context, next) => {
+        const availableTenants = await library.getAvailableTenants(context.auth.id);
+        if (
+          /**
+           * Authenticated user with `ManageTenant` scope can update any user tenant;
+           * authenticated user with `ManageTenantSelf` scope can only update their own tenant.
```
Comment (simeng-li) [inline packages/cloud/src/routes/tenants.ts:23]: Align the capitalization?   Besides, feels like the comments could leave at the enum scope definition place.

File: packages/cloud/src/routes/tenants.ts [hunk 2]
```diff
@@ -9,6 +9,37 @@ export const tenantsRoutes = (library: TenantsLibrary) =>
...
+           */
+          !(
+            context.auth.scopes.includes(CloudScope.ManageTenant) ||
+            (context.auth.scopes.includes(CloudScope.ManageTenantSelf) &&
+              availableTenants.map(({ id }) => id).includes(context.guarded.params.tenantId))
```
Comment (simeng-li) [inline packages/cloud/src/routes/tenants.ts:28]: no available tenants found should not return 403.

File: packages/cloud/src/libraries/tenants.ts [hunk 3]
```diff
@@ -67,6 +68,12 @@ export class TenantsLibrary {
     });
   }
 
+  async updateTenantById(tenantId: string, payload: PatchTenant): Promise<TenantInfo> {
+    const { id, name, tag } = await this.queries.tenants.updateTenantById(tenantId, payload);
+
+    return { id, name, tag, indicator: getManagementApiResourceIndicator(id) };
+  }
```
Comment (simeng-li) [inline packages/cloud/src/libraries/tenants.ts:75]: nit:

```js
async updateTenantById(tenantId: string, payload: PatchTenant): Promise<TenantInfo> => 
({
  ...await this.queries.tenants.updateTenantById(tenantId, payload)
  indicator: getManagementApiResourceIndicator(id
})
```

File: packages/cloud/src/routes/tenants.ts [hunk 4]
```diff
@@ -13,6 +13,43 @@ export const tenantsRoutes = (library: TenantsLibrary) =>
...
+        /** Should throw 404 when users with `ManageTenantSelf` scope are attempting to change an unavailable tenant. */
+        const availableTenants = await library.getAvailableTenants(context.auth.id);
+        if (
+          !(
+            context.auth.scopes.includes(CloudScope.ManageTenant) ||
+            availableTenants.map(({ id }) => id).includes(context.guarded.params.tenantId)
+          )
+        ) {
+          throw new RequestError(`Can not find tenant ${context.guarded.params.tenantId}.`, 404);
+        }
```
Comment (simeng-li) [inline packages/cloud/src/routes/tenants.ts:41]: query availableTenants only if not CloudScope.ManageTenant?

```js
if (!context.auth.scopes.includes(CloudScope.ManageTenant)) {
  const availableTenants = await library.getAvailableTenants(context.auth.id);
  assertThat(availableTenants.map(({ id }) => id).includes(context.guarded.params.tenantId), new RequestError(`Can not find tenant ${context.guarded.params.tenantId}.`, 404) )
}

```

File: packages/cloud/src/libraries/tenants.ts (hunk 3 above)
Comment (darcyYe) [reply packages/cloud/src/libraries/tenants.ts:78]: The returning result of tenant `updateTenantById` query also returns `dbUser` and `dbUserPassword` which can be used to build DB connection. We just pick id, name and tag for security consideration.

**Review by simeng-li — APPROVED** [review]
overall LGTM, please fix the ci

**Pull request merged**